### DISCUSSION TITLE
[Proposal] Allow the "new" operator to be overloaded

### DISCUSSION THREAD

**Discussion by CPosingies** [dotnet/csharplang#6743] [Language Ideas]
There is the "new" pattern, like in
```

class FooClass
{
    int foo;
    int oldFoo;

    EventHandler<FooEventArgs> FooChanged;

    FooClass()
    {
        Foo = 7;
    }

    Foo
    {
        get
        {
            return foo;
        }

        set
        {
            oldFoo = foo;
            foo = value;
            RaiseFooChanged();
        }
    }

    void RaiseFooChanged()
    {
        FooChanged?.Invoke(this, new FooEventArgs(foo, oldFoo));
    }
}

```
And there are DI/IoC containers that can handle all this, like the MEF and lookalikes. The thing is... If you've ever written such a container, I guess you are pretty familiar with AOP nightmares ^^ But not only as a coder of such a microkernel, also as a user of such a framework. Look at the snipped above. The whole bunch is nothing more like an Event-raising version of an int, kinda aware of its own history and informing its environment about changes of its value.

For a microkernel there is no way to intercept the (implicit) creation of the int object. Yes, I could cover each and every built-in type of the runtime or even recreate the types within new classes (which I do actually), but this is ridiculous ideed.

**Comment by iam3yal** [comment]
Even if the language allowed you to overload the `new` operator you couldn't provide your own implementation to existing types so what's the point?

**Comment by bernd5** [comment]
Why not using a source generator which creates all the ceremony?

**Reply by jnm2** [reply]
https://www.nuget.org/packages/StrongInject is extremely good.
